btclightclient/partial_merkle_tree.go

```diff
@@ -0,0 +1,215 @@
+		return nil, err
+	}
+
+	return pmk, nil
+}
+
+func (pmk *PartialMerkleTree) CalcTreeWidth(height uint32) uint {
+	return (pmk.numberTransactions + (1 << height) - 1) >> height
+}
+
+func (pmk *PartialMerkleTree) Height() uint32 {
+	var nHeight uint32 = 0
+	for pmk.CalcTreeWidth(nHeight) > 1 {
+		nHeight++
+	}
+	return nHeight
+}
+
+/// Port logic from gettxoutproof from bitcoin-core
+/// TODO: Make error handler more sense
+func (pmk *PartialMerkleTree) computeMerkleProofRecursive(height, pos uint32, nBitUsed, nHashUsed *uint32, txID *chainhash.Hash) (*MerkleProof, error) {
+	if int(*nBitUsed) >= len(pmk.vBits) {
+		return nil, errors.New("Error")
+	}
+
+	fParentOfMatch := pmk.vBits[*nBitUsed]
+	*nBitUsed = *nBitUsed + 1
+
+	if height == 0 || !fParentOfMatch {
+		if int(*nHashUsed) >= len(pmk.vHash) {
+			return nil, errors.New("error")
+		}
+
+		hash := pmk.vHash[*nHashUsed]
+		*nHashUsed++
+		if height == 0 && fParentOfMatch {
+			return &MerkleProof{
+				nodeValue:  *hash,
+				merklePath: []chainhash.Hash{*hash},
+				pos:        pos,
+			}, nil
+		}
+
+		return &MerkleProof{
+			nodeValue: *hash,
+			merklePath: []chainhash.Hash{},
+			pos: uint32(pmk.numberTransactions) + 1,
+		}, nil
+		
+	} else {
+		left, err := pmk.computeMerkleProofRecursive(height-1, pos*2, nBitUsed, nHashUsed, txID)
+		var right *MerkleProof
+		if err != nil {
+			return nil, err
+		}
+		if uint(pos*2+1) < pmk.CalcTreeWidth(height-1) {
+			right, err = pmk.computeMerkleProofRecursive(height-1, pos*2+1, nBitUsed, nHashUsed, txID)
+			if err != nil {
+				return nil, err
+			}
+
+			if left.nodeValue.IsEqual(&right.nodeValue) {
+				return nil, errors.New("error")
+			}
+		} else {
+			right = left
+		}
+
+		nodeValue := Hash256MerkleStepHashChain(&left.nodeValue, &right.nodeValue)
+		// Compute new proof
+		if left.pos != uint32(pmk.numberTransactions) {
+			// txID on the left side
+			return &MerkleProof {
+				nodeValue : *nodeValue,
+				pos: left.pos,
+				merklePath : append(left.merklePath, right.nodeValue),
+			}, nil
+		}
+		// TxID on right side
+		return &MerkleProof {
+				nodeValue : *nodeValue,
+				pos: right.pos,
+				merklePath : append(right.merklePath, left.nodeValue),
+		}, nil 
+	}
+}
+
+func (pmk *PartialMerkleTree) ComputeMerkleProof(txID string) (*MerkleProof, error) {
+	txIDHash, err := chainhash.NewHashFromStr(txID);
+	if err != nil {
+		return nil, err
+	}
+	height := pmk.Height();
+	nUsedBit := uint32(0)
+	nUsedHash := uint32(0)	
+	return pmk.computeMerkleProofRecursive(height, 0, &nUsedBit, &nUsedHash, txIDHash)
+}
+
+// Hash256MerkleStep concatenates and hashes two inputs for merkle proving
+func Hash256MerkleStep(a, b []byte) chainhash.Hash {
+	c := []byte{}
+	c = append(c, a...)
+	c = append(c, b...)
+	return  chainhash.DoubleHashH(c)
+}
+
+// TODO: make it more simple
+func Hash256MerkleStepHashChain(a, b *chainhash.Hash) *chainhash.Hash {
+	x := [32]byte(*a)
+	y := [32]byte(*b)
+	z := Hash256MerkleStep(x[:], y[:])
+	return &z
+}
```

Comment:
Let's merge the previous function with this one (and remove Hash256MerkleStep)

```suggestion
func HashNodes(a, b *chainhash.Hash) *chainhash.Hash {
	c := make([]byte{}, 0, chainhash.HashSize*2)
	c = append(c, a...)
	c = append(c, b...)
	return &chainhash.DoubleHashH(c)
}
```